rye/tests/test_publish.rs

```diff
@@ -0,0 +1,151 @@
+use crate::common::{fixtures_path, rye_cmd_snapshot, Space};
+
+mod common;
+
+// Test publish command
+#[test]
+fn test_publish() {
+    let space = Space::new();
+    space.init("test-project");
+
+    let fixtures_path = fixtures_path();
+    let wheel_path = fixtures_path.join("twine-1.5.0-py2.py3-none-any.whl");
+
+    let credentials_path = space.rye_home().join("credentials");
+    let credentials = std::fs::read_to_string(&credentials_path).unwrap();
+
+    // prompt for token
+    rye_cmd_snapshot!(space.rye_cmd().arg("publish").arg(&wheel_path), @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+    No access token found, generate one at: https://pypi.org/manage/account/token/
+
+    ----- stderr -----
+    Access token: error: IO error: not a terminal
+
+    Caused by:
+        not a terminal
+    "###);
+
+    rye_cmd_snapshot!(space.rye_cmd().arg("publish").arg("--skip-save-credentials").arg(&wheel_path), @r###"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+    No access token found, generate one at: https://pypi.org/manage/account/token/
+    Uploading distributions to https://upload.pypi.org/legacy/
+    Uploading twine-1.5.0-py2.py3-none-any.whl
+    [?25l
+[2K  0%                                          0.0/25.2 kB • --:-- • ?
+[2K  0%                                          0.0/25.2 kB • --:-- • ?
+[2K100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 25.2/25.2 kB • 00:00 • 64.7 MB/s
+[2K100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 25.2/25.2 kB • 00:00 • 64.7 MB/s
+    [?25hWARNING  Error during upload. Retry with the --verbose option for more details. 
+    ERROR    HTTPError: 403 Forbidden from https://upload.pypi.org/legacy/          
+             Invalid or non-existent authentication information. See                
+             https://pypi.org/help/#invalid-auth for more information.              
+
+    ----- stderr -----
+    Access token: error: failed to publish files
+    "###);
+
+    // invalid auth ('token')
+    rye_cmd_snapshot!(space.rye_cmd().arg("publish").arg("--skip-save-credentials").arg("--username").arg("user").arg("--token").arg("token").arg(&wheel_path), @r###"
```

Comment:
I probably won't end up running these tests with defaults to avoid unnecessarily interacting with PyPI.